File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_c1/structural_recursion_1.py
"""
Graph BFS - Structural Change Submission (queue -> recursion)
Same algorithm and output as organic_3.py (level-order BFS with distances),
but the iterative deque/while-loop is replaced with a recursive function
that processes one whole frontier (level) per call.
"""


def bfs_with_levels(graph, source):
    distance = {source: 0}
    order = []

    def process_frontier(frontier):
        if not frontier:
            return

        next_frontier = []
        for node in frontier:
            order.append((node, distance[node]))
            for neighbor in graph.get(node, []):
                if neighbor not in distance:
                    distance[neighbor] = distance[node] + 1
                    next_frontier.append(neighbor)

        process_frontier(next_frontier)

    process_frontier([source])
    return order


def print_levels(result):
    for node, level in result:
        print(f"Node {node} -> level {level}")


if __name__ == "__main__":
    graph = {
        1: [2, 3],
        2: [1, 4],
        3: [1, 4, 5],
        4: [2, 3, 6],
        5: [3, 6],
        6: [4, 5],
    }
    print_levels(bfs_with_levels(graph, 1))

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_c1/structural_recursion_1.java
/*
 * Graph BFS - Structural Change Submission (queue -> recursion)
 * Same algorithm and output as organic_3.java, but iterative while loop is
 * replaced with recursive function processing one level frontier per call.
 */

import java.util.*;

class Solution {
    public static List<String> bfsWithLevels(Map<Integer, List<Integer>> graph, int source) {
        Map<Integer, Integer> distance = new HashMap<>();
        List<String> order = new ArrayList<>();

        distance.put(source, 0);
        processFrontier(graph, Collections.singletonList(source), distance, order);
        return order;
    }

    private static void processFrontier(Map<Integer, List<Integer>> graph, List<Integer> frontier,
                                        Map<Integer, Integer> distance, List<String> order) {
        if (frontier.isEmpty()) {
            return;
        }

        List<Integer> nextFrontier = new ArrayList<>();
        for (int node : frontier) {
            order.add(node + ":" + distance.get(node));
            List<Integer> neighbors = graph.getOrDefault(node, Collections.emptyList());
            for (int neighbor : neighbors) {
                if (!distance.containsKey(neighbor)) {
                    distance.put(neighbor, distance.get(node) + 1);
                    nextFrontier.add(neighbor);
                }
            }
        }

        processFrontier(graph, nextFrontier, distance, order);
    }

    public static void main(String[] args) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(1, Arrays.asList(2, 3));
        graph.put(2, Arrays.asList(1, 4));
        graph.put(3, Arrays.asList(1, 4, 5));
        graph.put(4, Arrays.asList(2, 3, 6));
        graph.put(5, Arrays.asList(3, 6));
        graph.put(6, Arrays.asList(4, 5));

        List<String> res = bfsWithLevels(graph, 1);
        for (String item : res) {
            String[] parts = item.split(":");
            System.out.println("Node " + parts[0] + " -> level " + parts[1]);
        }
    }
}
